File: code/zato-web-admin/src/zato/admin/web/views/channel/file_transfer.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# Zato
from zato.admin.web.forms.channel.file_transfer import CreateForm, EditForm
from zato.admin.web.views import CreateEdit, Delete as _Delete, get_outconn_rest_list, Index as _Index
from zato.common.api import FILE_TRANSFER, GENERIC
from zato.common.json_internal import dumps
from zato.common.model.file_transfer import FileTransferChannel

# ################################################################################################################################
# ################################################################################################################################

source_type_ftp   = FILE_TRANSFER.SOURCE_TYPE.FTP.id
source_type_local = FILE_TRANSFER.SOURCE_TYPE.LOCAL.id
source_type_sftp  = FILE_TRANSFER.SOURCE_TYPE.SFTP.id

# ################################################################################################################################
# ################################################################################################################################

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'channel-file-transfer'
    template = 'zato/channel/file-transfer.html'
    service_name = 'zato.generic.connection.get-list'
    output_class = FileTransferChannel
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = 'cluster_id', 'type_'
        output_required = 'id', 'name', 'is_active', 'source_type', 'pickup_from_list'
        output_optional = 'service_list', 'topic_list', 'move_processed_to', 'file_patterns', 'parse_with', \
            'should_read_on_pickup', 'should_parse_on_pickup', 'should_delete_after_pickup', 'ftp_source_id', 'sftp_source_id', \
            'scheduler_job_id', 'ftp_source_name', 'sftp_source_name', 'is_case_sensitive', 'is_line_by_line', 'is_hot_deploy', \
            'binary_file_patterns', 'data_encoding', 'outconn_rest_list'
        output_repeated = True

# ################################################################################################################################

    def handle(self):
        return {
            'create_form': CreateForm(req=self.req),
            'edit_form': EditForm(prefix='edit', req=self.req),
        }

# ################################################################################################################################

    def on_before_append_item(self, item):
        # type: (FileTransferChannel) -> FileTransferChannel

        if item.service_list:
            item.service_list = item.service_list if isinstance(item.service_list, list) else [item.service_list]
            item.service_list = sorted(item.service_list)
            item.service_list_json = dumps(item.service_list)

        if item.topic_list:
            item.topic_list = item.topic_list if isinstance(item.topic_list, list) else [item.topic_list]
            item.topic_list = sorted(item.topic_list)
            item.topic_list_json = dumps(item.topic_list)

        if item.outconn_rest_list:

            # All REST outgoing connections for the cluster
            all_outconn_rest_list = get_outconn_rest_list(self.req)

            item.outconn_rest_list = item.outconn_rest_list if isinstance(item.outconn_rest_list, list) else \
                [item.outconn_rest_list]
            item.outconn_rest_list_by_name = sorted(all_outconn_rest_list[int(elem)] for elem in item.outconn_rest_list if elem)
            item.outconn_rest_list_json = dumps(item.outconn_rest_list)

        return item

# ################################################################################################################################
# ################################################################################################################################

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = 'name', 'is_active', 'source_type', 'pickup_from_list'
        input_optional = 'service_list', 'topic_list', 'move_processed_to', 'file_patterns', 'parse_with', 'should_read_on_pickup', \
            'should_parse_on_pickup', 'should_delete_after_pickup', 'ftp_source_id', 'sftp_source_id', 'scheduler_job_id', \
            'is_case_sensitive', 'is_line_by_line', 'is_hot_deploy', 'binary_file_patterns', 'data_encoding', \
            'outconn_rest_list'
        output_required = 'id', 'name'

    def populate_initial_input_dict(self, initial_input_dict):
        initial_input_dict['type_'] = GENERIC.CONNECTION.TYPE.CHANNEL_FILE_TRANSFER
        initial_input_dict['is_internal'] = False
        initial_input_dict['is_channel'] = True
        initial_input_dict['is_outconn'] = False
        initial_input_dict['sec_use_rbac'] = False
        initial_input_dict['pool_size'] = 1

    def pre_process_item(self, name, value):
        if name in ('service_list', 'topic_list', 'outconn_rest_list'):
            if value:
                if isinstance(value, list):
                    value = sorted({elem for elem in value if elem})
                else:
                    value = [value] if value else []

        return value

    def post_process_return_data(self, return_data):

        service_list_html = ''
        topic_list_html   = ''
        outconn_rest_list_html   = ''

        cluster_id = self.input_dict['cluster_id']
        service_list = sorted(set(self.input_dict['service_list'] or []))
        topic_list = sorted(set(self.input_dict['topic_list'] or []))
        outconn_rest_list = self.input_dict['outconn_rest_list'] or []

        if outconn_rest_list:

            # All REST outgoing connections for the cluster
            all_outconn_rest_list = get_outconn_rest_list(self.req)

            outconn_rest_list = outconn_rest_list if isinstance(outconn_rest_list, list) else [outconn_rest_list]
            outconn_rest_list = sorted({all_outconn_rest_list[int(elem)] for elem in outconn_rest_list if elem})

        #
        # Services
        #

        for service_name in service_list:

            if not service_name:
                continue

            # The list of services as HTML
            service_list_html += """
            <span class="form_hint">S</span>→
                <a href="/zato/service/overview/{service_name}/?cluster={cluster_id}">{service_name}</a>
            <br/>
            """.format(**{
                'service_name': service_name,
                'cluster_id': cluster_id
            })

        #
        # Topics
        #

        for topic_name in topic_list:

            if not topic_name:
                continue

            # The list of topics as HTML
            topic_list_html += """
            <span class="form_hint">T</span>→
                <a href="/zato/pubsub/topic/?cluster={cluster_id}&amp;query={topic_name}">{topic_name}</a>
            <br/>
            """.format(**{
                'topic_name': topic_name,
                'cluster_id': cluster_id
            })

        #
        # REST
        #

        for outconn_name in outconn_rest_list:

            if not outconn_name:
                continue

            # The list of REST outconns as HTML
            outconn_rest_list_html += """
            <span class="form_hint">R</span>→
                <a href="/zato/http-soap/?cluster={cluster_id}&amp;connection=outgoing&amp;transport=plain_http&amp;query={outconn_name}">{outconn_name}</a>
            <br/>
            """.format(**{
                'outconn_name': outconn_name,
                'cluster_id': cluster_id
            })

        # Get human-readable names of the source for this channel
        if self.input.source_type == source_type_local:
            # Nothing to do in this case
            pass

        elif self.input.source_type == source_type_ftp:

            if self.input.ftp_source_id:
                response = self.req.zato.client.invoke('zato.outgoing.ftp.get-by-id', {
                    'cluster_id': cluster_id,
                    'id': self.input.ftp_source_id,
                })

                return_data['ftp_source_name'] = response.data['name']

        elif self.input.source_type == source_type_sftp:
            raise NotImplementedError()

        else:
            raise ValueError('Unknown self.input.source_type')

        #
        # Return data
        #

        return_data['recipients_html'] = service_list_html + topic_list_html + outconn_rest_list_html
        return_data['service_list'] = service_list
        return_data['topic_list'] = topic_list
        return_data['outconn_rest_list'] = outconn_rest_list

    def success_message(self, item):
        return 'Successfully {} file transfer channel `{}`'.format(self.verb, item.name)

# ################################################################################################################################
# ################################################################################################################################

class Create(_CreateEdit):
    url_name = 'channel-file-transfer-create'
    service_name = 'zato.generic.connection.create'

# ################################################################################################################################
# ################################################################################################################################

class Edit(_CreateEdit):
    url_name = 'channel-file-transfer-edit'
    form_prefix = 'edit-'
    service_name = 'zato.generic.connection.edit'

# ################################################################################################################################
# ################################################################################################################################

class Delete(_Delete):
    url_name = 'channel-file-transfer-delete'
    error_message = 'Could not delete file transfer channel'
    service_name = 'zato.generic.connection.delete'

# ################################################################################################################################
# ################################################################################################################################

File: code/zato-web-admin/src/zato/admin/static/js/channel/file-transfer.js

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.ChannelFileTransfer = new Class({
    toString: function() {
        var s = '<ChannelFileTransfer id:{0} name:{1} is_active:{2}>';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.ChannelFileTransfer;
    $.fn.zato.data_table.new_row_func = $.fn.zato.channel.file_transfer.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'source_type', 'pickup_from_list', 'file_patterns']);

    $('#id_source_type').change(function() {
        $.fn.zato.channel.file_transfer.on_source_type_changed('');
    });

    $('#id_edit-source_type').change(function() {
        $.fn.zato.channel.file_transfer.on_source_type_changed('edit-');
    });

})

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.channel.file_transfer.on_source_type_changed = function(prefix) {

    let id_prefix = '#id_'+ prefix;
    let source_type = $(id_prefix + 'source_type').val();

    if(source_type) {

        let ftp       = $(id_prefix + 'ftp_source_id');
        let sftp      = $(id_prefix + 'sftp_source_id');

        let scheduler_create_row = $('#tr_create_scheduler_job_id');
        let scheduler_edit_row = $('#tr_edit_scheduler_job_id');

        if(source_type == 'local') {

            scheduler_create_row.addClass('hidden');
            scheduler_edit_row.addClass('hidden');

            ftp.addClass('hidden');
            sftp.addClass('hidden');
        }

        else if(source_type == 'ftp') {

            scheduler_create_row.removeClass('hidden');
            scheduler_edit_row.removeClass('hidden');

            ftp.removeClass('hidden');
            sftp.addClass('hidden');
        }

        else if(source_type == 'sftp') {

            scheduler_create_row.removeClass('hidden');
            scheduler_edit_row.removeClass('hidden');

            sftp.removeClass('hidden');
            ftp.addClass('hidden');
        }
    }
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.channel.file_transfer.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new file transfer channel', null);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.channel.file_transfer.edit = function(id) {

    let instance = $.fn.zato.data_table.data[id];
    let source_type = $('#id_edit-'+ instance.source_type +'_source_id');
    source_type.removeClass('hidden');

    if(instance.source_type != "local") {

        let scheduler_create_row = $('#tr_create_scheduler_job_id');
        let scheduler_edit_row = $('#tr_edit_scheduler_job_id');

        scheduler_create_row.removeClass('hidden');
        scheduler_edit_row.removeClass('hidden');
    };

    $('#id_edit-ftp_source_name').val(instance.ftp_source_name);
    $('#id_edit-sftp_source_name').val(instance.sftp_source_name);

    $.fn.zato.data_table.multirow.remove_multirow_added();
    $.fn.zato.data_table.edit('edit', 'Update the file transfer channel', id, false);

    $.fn.zato.data_table.multirow.populate_field('service_list', instance.service_list);
    $.fn.zato.data_table.multirow.populate_field('topic_list', instance.topic_list);
    $.fn.zato.data_table.multirow.populate_field('outconn_rest_list', instance.outconn_rest_list);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.channel.file_transfer.data_table.new_row = function(item, data, include_tr) {
    let row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    let is_active = item.is_active == true;
    let cluster_id = $(document).getUrlParam('cluster');

    let source_html = $.fn.zato.empty_value;
    let pickup_from_list_html = '';
    let recipients_html = data.recipients_html || '<span class="form_hint">---</span>';

    if(item.source_type == 'local') {
        source_html = 'Local';
    }
    else if(item.source_type == 'ftp') {
        if(item.ftp_source_name) {
            source_html = `<a href="/zato/outgoing/ftp/?cluster=${item.cluster_id}&amp;query=${item.ftp_source_name}">${item.ftp_source_name}</a>`;
        }
        else {
            source_html = 'FTP <span class="form_hint"><br/>(No connection assigned)</span>';
        }
    }
    else if(item.source_type == 'sftp') {
        if(item.sftp_source_name) {
            source_html = `<a href="/zato/outgoing/sftp/?cluster=${item.cluster_id}&amp;type_=outconn-sftp&amp;query=${item.sftp_source_name}">${item.sftp_source_name}</a>`;
        }
        else {
            source_html = 'SFTP <span class="form_hint"><br/>(No connection assigned)</span>';
        }
    }

    pickup_from_list_html += String.replace(item.pickup_from_list.replace("\n", "<br/>"));
    pickup_from_list_html += '<br/>';
    pickup_from_list_html += item.file_patterns;

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";

    // 1
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', source_html);

    // 2
    row += String.format('<td>{0}</td>', pickup_from_list_html);
    row += String.format('<td>{0}</td>', item.move_processed_to ? item.move_processed_to : $.fn.zato.empty_value);

    // 3
    row += String.format('<td>{0}</td>', recipients_html);
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.channel.file_transfer.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.channel.file_transfer.delete_({0});'>Delete</a>", item.id));

    // 4
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", item.is_active);

    // 5
    row += String.format("<td class='ignore'>{0}</td>", item.should_read_on_pickup);
    row += String.format("<td class='ignore'>{0}</td>", data.service_list);
    row += String.format("<td class='ignore'>{0}</td>", data.topic_list);

    // 6
    row += String.format("<td class='ignore'>{0}</td>", item.should_parse_on_pickup);
    row += String.format("<td class='ignore'>{0}</td>", item.parse_with);
    row += String.format("<td class='ignore'>{0}</td>", item.should_delete_after_pickup);

    // 7
    row += String.format("<td class='ignore'>{0}</td>", item.is_internal);
    row += String.format("<td class='ignore'>{0}</td>", item.source_type);
    row += String.format("<td class='ignore'>{0}</td>", item.ftp_source_id);

    // 8
    row += String.format("<td class='ignore'>{0}</td>", item.sftp_source_id);
    row += String.format("<td class='ignore'>{0}</td>", item.file_patterns);
    row += String.format("<td class='ignore'>{0}</td>", item.scheduler_job_id);

    // 9
    row += String.format("<td class='ignore'>{0}</td>", item.is_line_by_line);

    // 10
    row += String.format("<td class='ignore'>{0}</td>", item.pickup_from_list);
    row += String.format("<td class='ignore'>{0}</td>", item.ftp_source_name);
    row += String.format("<td class='ignore'>{0}</td>", item.sftp_source_name);

    // 11
    row += String.format("<td class='ignore'>{0}</td>", item.is_case_sensitive);
    row += String.format("<td class='ignore'>{0}</td>", item.move_processed_to);
    row += String.format("<td class='ignore'>{0}</td>", item.is_hot_deploy);

    // 12
    row += String.format("<td class='ignore'>{0}</td>", item.binary_file_patterns);
    row += String.format("<td class='ignore'>{0}</td>", item.data_encoding);
    row += String.format("<td class='ignore'>{0}</td>", data.outconn_rest_list);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.channel.file_transfer.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'File transfer channel `{0}` deleted',
        'Are you sure you want to delete file transfer channel `{0}`?',
        true);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
